### ISSUE TITLE
Client resiliency to slow replicas part 1

### ISSUE THREAD

**Issue by j82w** [Azure/azure-cosmos-dotnet-v3#2221]
### Problem:
The current replica selection is always doing a random selection then waiting for all the results. This means that if a replica is slow the SDK will wait until it get's a response or timeouts.

For example strong consistency: 
1. Send 2 requests in parallel and wait for both requests to complete
2. If one the request is slow it waits until it gets a response or timeouts (default 5 seconds). 
3. Another replica is selected and a new request will be sent.
4. New request succeeds, but will cause a slow request since the original can take up to 5 seconds to timeout.

```
Request -> A --|
        -> B ------------------|
                                -> C --|
                                        -> Compute result
```

### Solution part 1:
Instead of waiting for all request to complete a additional request should be sent while waiting for the slow request.

1. Add a StoreReader latency tracker which will use a concurrent dictionary to track latency of request by operation type, and region. Operation type is included because aggregate operations like query will take significantly longer than point operations. This doesn't mean the request was slow.
2. Send a request
3. If the request take more than 25% above the average latency a duplicate request will be sent
4. The SDK will wait for one of the request to complete and use the result of the request that finished first.
5. The slower request will use a continue with to log any exception and to let the request complete. This avoid additional over head cause by using a cancellation token and throwing an exception.

```
Request -> A --|     
        -> B ------------------|
                ->C --|
                       -> Compute result
```

### Solution part 2:
Will be done in a separate issue and will be adding waits to the replicas selection to help avoid picking slow replica in future request.

**Comment by ealsur** [comment]
We need to keep in mind if any solution that prioritizes latency might cause an RU increase because of extra requests being sent when the original one, albeit slow, succeeds.

**Comment by j82w** [comment]
That's a good point. We can add a feature flag to allow user to control it. This should be minimized with part 2 by avoiding the slow partitions. @kirankumarkolli  and @FabianMeiswinkel what are your thoughts?
